A67-xds-dynamic-parameters.md

```diff
@@ -0,0 +1,246 @@
+of a dynamic parameter could be `{key: "env", value: "prod"}`.
+
+We will add support for dynamic parameters in gRPC in all languages. For this,
+we will extended the bootstrap JSON, and modify how ADS requests are constructed
+and how ADS responses are parsed.
+
+## Background
+
+[TP2](https://github.com/cncf/xds/blob/main/proposals/TP2-dynamically-generated-cacheable-xds-resources.md)
+describes in detail how dynamic parameters will function. While TP2 discusses
+the functionality in terms of xDS generally, this document instead goes into
+details about what more specifically in gRPC will need to be modified.
+
+This document assumes gRPC's `XdsClient` is only used by leaf clients. If at
+some point gRPC's `XdsClient` gets reused in an xDS proxy, additional
+functionality for processing dynamic parameter constraints will need to be
+added.
+
+TP2 suggests that:
+
+-   Subscriptions to xdstp named resources allow for a separate set of dynamic
+    parameters per authority.
+-   Subscriptions to non-xdstp named resources use a set of dynamic parameters
+    configured globally for the client.
+
+We will follow this suggestion.
+
+### Request
+
+Currently a request that basically look like this:
+
+```text
+resource_names: "xdstp://xds.example.com/envoy.config.listener.v3.Listener/foo"
+type_url: "type.googleapis.com/envoy.config.listener.v3.Listener"
+```
+
+With dynamic parameters it will look like this:
+
+```text
+resource_locators {
+  name: "xdstp://xds.example.com/envoy.config.listener.v3.Listener/foo"
+  dynamic_parameters: [ {key: "env", value: "prod"} ]
+}
+type_url: "type.googleapis.com/envoy.config.listener.v3.Listener"
+```
+
+### Response
+
+TP2 specifies that xDS servers responding to requests containing dynamic
+parameters should wrap returned resources in Resource messages, and use the
+`resource_name` field instead of the name field.
+
+When we make a request with `resource_names` as we currently do, we expect to
+get a response like this:
+
+```text
+type_url: "type.googleapis.com/envoy.config.listener.v3.Listener"
+resources {
+  [type.googleapis.com/envoy.config.listener.v3.Listener] {
+    name: "xdstp://xds.example.com/envoy.config.listener.v3.Listener/foo"
+    ...
+  }
+}
+```
+
+With dynamic parameters, we should instead be getting a response with a
+Resource-wrapper like this:
+
+```text
+type_url: "type.googleapis.com/envoy.config.listener.v3.Listener"
+resources {
+  [type.googleapis.com/envoy.service.discovery.v3.Resource] {
+    resource_name {
+      name: "xdstp://xds.example.com/envoy.config.listener.v3.Listener/foo"
+      dynamic_parameter_constraints {
+        constraint { key: "env", value: "prod" }
+      }
+    }
+    resource: {
+      [type.googleapis.com/envoy.config.listener.v3.Listener] {
+        name: "xdstp://xds.example.com/envoy.config.listener.v3.Listener/foo"
+        ...
+      }
+    }
+  }
+}
+```
+
+### Related Proposals:
```

Comment:
It's probably also relevant to add a link to [gRFC A47: xDS Federation](https://github.com/grpc/proposal/blob/master/A47-xds-federation.md), which describes the common structure of the XdsClient.

Reply:
Added.